clang/include/clang/CConv/ConstraintsGraph.h

```diff
@@ -12,125 +12,269 @@
+
+private:
+  DataType Data;
+
+  llvm::SetVector<EdgeType *> PredEdges;
+};
+
+template<typename Data, typename EdgeType>
+struct llvm::GraphTraits<DataNode<Data, EdgeType> *> {
+  using NodeRef = DataNode<Data, EdgeType> *;
+
+  static NodeRef GetTargetNode(EdgeType *P) {
+    return &P->getTargetNode();
+  }
+
+  using ChildIteratorType = mapped_iterator
+      <typename DataNode<Data, EdgeType>::iterator, decltype(&GetTargetNode)>;
+
+  static NodeRef getEntryNode(NodeRef N) { return N; }
+  static ChildIteratorType child_begin(NodeRef N) {
+    return ChildIteratorType(N->begin(), &GetTargetNode);
   }
+  static ChildIteratorType child_end(NodeRef N) {
+    return ChildIteratorType(N->end(), &GetTargetNode);
+  }
+};
+
+template<class DataType>
+struct DataEdge : public llvm::DGEdge<DataNode<DataType>, DataEdge<DataType>> {
+  typedef llvm::DGEdge<DataNode<DataType>, DataEdge<DataType>> SuperType;
+  explicit DataEdge(DataNode<DataType> &Node) : SuperType(Node) {}
+  DataEdge(const DataEdge &E) : SuperType(E) {}
 };
 
-class ConstraintsGraph
-    : public BaseGraph<boost::adjacency_list<boost::setS, boost::vecS,
-                                             boost::bidirectionalS, Atom *>> {
+class GraphVizOutputGraph;
+
+// Define a general purpose extension to the llvm provided graph class that
+// stores some data at each node in the graph. This is used by the checked and
+// pointer type constraint graphs (which store atoms at each node) as well as
+// the array bounds graph (which stores BoundsKeys).
+template<typename Data, typename EdgeType = DataEdge<Data>>
+class DataGraph :
+    public llvm::DirectedGraph<DataNode<Data, EdgeType>, EdgeType> {
 public:
-  typedef boost::adjacency_list<boost::setS, boost::vecS,
-                                boost::bidirectionalS, Atom*> DirectedGraphType;
+  typedef DataNode<Data, EdgeType> NodeType;
 
-  ConstraintsGraph() {
-    AllConstAtoms.clear();
-    AtomToVDMap.clear();
+  virtual ~DataGraph() {
+    for (auto *N : this->Nodes) {
+      for (auto *E : N->getEdges())
+        delete E;
+      N->getEdges().clear();
+      delete N;
+      N = nullptr;
+    }
+    this->Nodes.clear();
   }
-  virtual ~ConstraintsGraph() {
-    AllConstAtoms.clear();
-    AtomToVDMap.clear();
+
+  void removeEdge(Data Src, Data Dst) {
+    auto *NSrc = this->findNode(NodeType(Src));
+    auto *NDst = this->findNode(NodeType(Dst));
+    assert(NSrc != this->end() && NDst != this->end());
+    llvm::SmallVector<EdgeType*, 10> Edges;
+    (*NDst)->findEdgesTo(**NSrc, Edges);
+    for (EdgeType *E : Edges) {
+      (*NDst)->removeEdge(*E);
+      delete E;
+    }
   }
 
-  void addConstraint(Geq *C, const Constraints &CS);
+  void addEdge(Data L, Data R) {
+    NodeType *BL = this->addVertex(L);
+    NodeType *BR = this->addVertex(R);
 
+    auto *BLR = new EdgeType(*BR);
+    BL->addEdge(*BLR);
+    auto *BRL = new EdgeType(*BL);
+    BR->addPred(*BRL);
+  }
 
```

Comment:
In general I find all of this code very hard to understand. Ultimately we are talking about a pretty simple data structure, but that simplicity is smothered by all of the boilerplate. Is there no better way?

Reply:
All of the the `GraphTraits` template specialization boilerplate is required to use the llvm provided graphviz generation and breadth first search implementation.

Reply:
I'm overall disappointed with the` llvm::DirectedGraph` class and other utilities. It's not a very nice set of tools to work with.

Reply:
OK.